plugins/org.komodo.core/src/org/komodo/repository/search/ObjectSearcher.java

```diff
@@ -0,0 +1,433 @@
+
+    private String customWhereClause;
+
+    /**
+     * @param repository the repository to search
+     */
+    public ObjectSearcher(Repository repository) {
+        ArgCheck.isNotNull(repository);
+        this.repository = repository;
+    }
+
+    private Session getSession(final UnitOfWork transaction) {
+        return ((UnitOfWorkImpl)transaction).getSession();
+    }
+
+    private boolean isEmpty(Collection<?> list) {
+        return list == null || list.isEmpty();
+    }
+
+    /**
+     * @return the repository
+     */
+    public Repository getRepository() {
+        return this.repository;
+    }
+
+    /**
+     * Add a node type to the From clause
+     *
+     * @param type the type to be added
+     * @param alias the alias of the type
+     */
+    public void addFromType(String type, String alias) {
+        ArgCheck.isNotEmpty(type);
+        String failMsg = "From clause data should be added prior to where clause data"; //$NON-NLS-1$
+        ArgCheck.isTrue(whereClauses == null, failMsg);
+        ArgCheck.isTrue(customWhereClause == null, failMsg);
+
+        if (fromTypes == null)
+            fromTypes = new LinkedHashSet<FromType>();
+
+        fromTypes.add(new FromType(type, alias));
+    }
+
+    /**
+     * Add a node type to the From clause
+     *
+     * @param type the type to be added
+     */
+    public void addFromType(String type) {
+        addFromType(type, null);
+    }
+
+    /**
+     * @return set of node types for the From clause
+     */
+    private Set<FromType> getFromTypes() {
+        if (fromTypes == null || fromTypes.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        return fromTypes;
+    }
+
+    private String checkWhereAlias(String alias) {
+        //
+        // Check that where type alias is valid for set of from types
+        //
+        if (getFromTypes().size() == 1 && StringUtils.isEmpty(alias)) {
+            // Only 1 from type and alias is empty so assume alias of single from type
+            alias = getFromTypes().iterator().next().getAlias();
+        } else {
+            //
+            // More than 1 from type or alias is not empty
+            //
+            ArgCheck.isNotEmpty(alias);
+
+            boolean aliasTypeFound = false;
+            for (FromType fromType : getFromTypes()) {
+                if (fromType.getAlias().equals(alias)) {
+                    aliasTypeFound = true;
+                    break;
+                }
+            }
+
+            ArgCheck.isTrue(aliasTypeFound, "Where clause alias is unknown to from clause"); //$NON-NLS-1$
+        }
+
+        return alias;
+    }
+
+    private SetClause findWhereSetClause(String alias, String property) {
+        for (Clause clause : whereClauses) {
+            if (! (clause instanceof SetClause))
+                continue;
+
+            SetClause setClause = (SetClause) clause;
+            if (setClause.getAlias().equalsIgnoreCase(alias) &&
```

Comment:
Not if the SetClause was valid. If using as set clause, ie.
SELECT * FROM [blah] n1 where n1.name IN ['bob', 'james']
the alias must exist as well as the property.